apps/emqx_retainer/src/emqx_retainer_mnesia.erl

```diff
@@ -55,201 +90,289 @@ create_resource(#{storage_type := StorageType}) ->
 
 clear_expired(_) ->
     NowMs = erlang:system_time(millisecond),
-    MsHd = #retained{topic = '$1', msg = '_', expiry_time = '$3'},
-    Ms = [{MsHd, [{'=/=', '$3', 0}, {'<', '$3', NowMs}], ['$1']}],
+    QH = qlc:q([ TopicTokens || #retained_message{
+                                   topic = TopicTokens,
+                                   expiry_time = ExpiryTime
+                                  } <- mnesia:table(?TAB_MESSAGE, [{lock, write}]),
+                                (ExpiryTime =/= 0) and (ExpiryTime < NowMs)]),
     Fun = fun() ->
-                  Keys = mnesia:select(?TAB, Ms, write),
-                  lists:foreach(fun(Key) -> mnesia:delete({?TAB, Key}) end, Keys)
+                  QC = qlc:cursor(QH),
+                  clear_batch(db_indices(write), QC)
           end,
     {atomic, _} = mria:transaction(?RETAINER_SHARD, Fun),
     ok.
 
 delete_message(_, Topic) ->
-    case emqx_topic:wildcard(Topic) of
-        true -> match_delete_messages(Topic);
-        false ->
-            Tokens = topic2tokens(Topic),
-            Fun = fun() ->
-                       mnesia:delete({?TAB, Tokens})
-                  end,
-            _ = mria:transaction(?RETAINER_SHARD, Fun),
-            ok
-    end,
+    Tokens = topic_to_tokens(Topic),
+    DeleteFun = case emqx_topic:wildcard(Topic) of
+                    false ->
+                        fun() ->
+                                ok = delete_message_by_topic(Tokens, db_indices(write))
+                        end;
+                    true ->
+                        fun() ->
+                                QH = topic_search_table(Tokens),
+                                qlc:fold(
+                                  fun(TopicTokens, _) ->
+                                          ok = delete_message_by_topic(TopicTokens, db_indices(write))
+                                  end,
+                                  undefined,
+                                  QH)
+                        end
+                end,
+    {atomic, _} = mria:transaction(?RETAINER_SHARD, DeleteFun),
     ok.
 
 read_message(_, Topic) ->
     {ok, read_messages(Topic)}.
 
-page_read(_, Topic, Page, Limit) ->
-    Cursor = make_cursor(Topic),
-    case Page > 1 of
-        true ->
-            _ = qlc:next_answers(Cursor, (Page - 1) * Limit),
-            ok;
-        _ ->
-            ok
-    end,
-    Rows = qlc:next_answers(Cursor, Limit),
-    qlc:delete_cursor(Cursor),
-    {ok, Rows}.
-
-match_messages(_, Topic, Cursor) ->
-    BatchReadNum = emqx:get_config([retainer, flow_control, batch_read_number]),
-    case Cursor of
-        undefined ->
-            case BatchReadNum of
-                0 ->
-                    {ok, sort_retained(match_messages(Topic)), undefined};
-                _ ->
-                    start_batch_read(Topic, BatchReadNum)
-            end;
-        _ ->
-            batch_read_messages(Cursor, BatchReadNum)
+match_messages(_, Topic, undefined) ->
```

Comment:
why isn't it a `/2` function ?

Reply:
It's a callback implementation; alternative retainers are supposed to be stateful, but mnesia-based keeps data in mnesia.